File: modules/front/src/common/api/services/dealService.ts
import { createResourceService } from "../module/service/resourceFactory";
import type {
  CreateDealDTO,
  DealDTO,
  MyPaymentsDTO,
  StatutDealType,
} from "../types";
import type { PageResponse, PaginationParams } from "@/common/types/pagination";
import { apiClient } from "./apiClient";

// Service de base avec méthodes CRUD standard
export const dealBaseService = createResourceService<DealDTO>(
  apiClient,
  "/deals",
);

// Service étendu avec méthodes spécifiques
export const dealService = {
  ...dealBaseService,

  create: (deal: CreateDealDTO) =>
    apiClient.post<DealDTO>("/deals", {
      body: deal,
    }),

  // Récupérer tous les deals (non paginé - pour rétrocompatibilité)
  getAll: () => apiClient.get<DealDTO[]>("/deals"),

  // Récupérer tous les deals avec pagination
  getAllPaginated: (params: PaginationParams) =>
    apiClient.get<PageResponse<DealDTO>>("/deals", { queryParams: params }),

  // Récupérer les deals par statut (non paginé - pour rétrocompatibilité)
  getByStatut: (statut: StatutDealType) =>
    apiClient.get<DealDTO[]>(`/deals/statut/${statut}`),

  // Récupérer les deals par statut avec pagination
  getByStatutPaginated: (statut: StatutDealType, params: PaginationParams) =>
    apiClient.get<PageResponse<DealDTO>>(`/deals/statut/${statut}`, {
      queryParams: params
    }),

  // Récupérer les deals par créateur (non paginé - pour rétrocompatibilité)
  getByCreateur: (createurUuid: string) =>
    apiClient.get<DealDTO[]>(`/deals/createur/${createurUuid}`),

  // Récupérer les deals par créateur avec pagination
  getByCreateurPaginated: (createurUuid: string, params: PaginationParams) =>
    apiClient.get<PageResponse<DealDTO>>(`/deals/createur/${createurUuid}`, {
      queryParams: params,
    }),

  // Récupérer les deals par catégorie (non paginé - pour rétrocompatibilité)
  getByCategorie: (categorieUuid: string) =>
    apiClient.get<DealDTO[]>(`/deals/categorie/${categorieUuid}`),

  // Récupérer les deals par catégorie avec pagination
  getByCategoriePaginated: (categorieUuid: string, params: PaginationParams) =>
    apiClient.get<PageResponse<DealDTO>>(`/deals/categorie/${categorieUuid}`, {
      queryParams: params,
    }),

  // Confirmer l'upload d'une image
  confirmImageUpload: (dealUuid: string, imageUuid: string) =>
    apiClient.patch<void>(`/deals/${dealUuid}/images/${imageUuid}/confirm`),

  // Obtenir l'URL de lecture d'une image
  getImageUrl: (dealUuid: string, imageUuid: string) =>
    apiClient.get<{ url: string }>(
      `/deals/${dealUuid}/images/${imageUuid}/url`,
    ),

  // Récupérer les villes disponibles
  getVilles: () => apiClient.get<string[]>("/deals/villes"),

  // Mettre à jour le statut d'un deal
  updateStatus: (uuid: string, statut: string) =>
    apiClient.patch<DealDTO>(`/deals/${uuid}/statut`, {
      body: { statut },
    }),

  // Mettre à jour les images d'un deal
  updateImages: (uuid: string, images: any) =>
    apiClient.patch<DealDTO>(`/deals/${uuid}/images`, {
      body: images,
    }),

  getParticipants: (dealUuid: string) =>
    apiClient.get<any[]>(`/deals/${dealUuid}/participants`),

  myPayments: () => apiClient.get<MyPaymentsDTO[]>("/deals/mes-paiements"),

  toggleFavoris: (uuid: string) =>
    apiClient.patch<DealDTO>(`/deals/${uuid}/favoris`),
};

File: modules/bff/bff-core/src/main/java/com/ulr/paytogether/core/domaine/service/DealService.java
package com.ulr.paytogether.core.domaine.service;

import com.ulr.paytogether.core.enumeration.StatutImage;
import com.ulr.paytogether.core.modele.DealModele;
import com.ulr.paytogether.core.enumeration.StatutDeal;

import java.util.List;
import java.util.Optional;
import java.util.Set;
import java.util.UUID;

/**
 * Interface du service Deal
 */
public interface DealService {

    /**
     * Créer un deal
     */
    DealModele creer(DealModele deal);

    /**
     * Lire un deal par son UUID
     */
    Optional<DealModele> lireParUuid(UUID uuid);

    /**
     * Lire tous les deals
     */
    List<DealModele> lireTous();

    /**
     * Lire les deals par statut
     */
    List<DealModele> lireParStatut(StatutDeal statut);

    /**
     * Lire les deals d'un créateur
     */
    List<DealModele> lireParCreateur(UUID createurUuid);

    /**
     * Lire les deals d'une catégorie
     */
    List<DealModele> lireParCategorie(UUID categorieUuid);

    /**
     * Mettre à jour un deal
     */
    DealModele mettreAJour(UUID uuid, DealModele deal);

    /**
     * Mettre à jour uniquement le statut d'un deal
     */
    DealModele mettreAJourStatut(UUID uuid, StatutDeal statut);

    /**
     * Mettre à jour uniquement les images d'un deal
     */
    DealModele mettreAJourImages(UUID uuid, DealModele deal);

    /**
     * Supprimer un deal par son UUID
     */
    void supprimerParUuid(UUID uuid);

    /**
     * Récupérer les villes disponibles pour les deals
     */
    Set<String> lireVillesDisponibles();

    /**
     * Mettre à jour le statut d'une image d'un deal
     */
    void mettreAJourStatutImage(UUID dealUuid, UUID imageUuid, StatutImage statut);

    /**
     * Obtenir l'URL de lecture d'une image d'un deal
     */
    String obtenirUrlLectureImage(UUID dealUuid, UUID imageUuid);
}
